File: Project5/Solutions/python/mdp.py
""" Problem 3 Markov Decision Process """

import csv
import copy
import math

class MDP:
    """ MDP Value-Iteration Algorithm Class """
    def __init__(self):
        # Local Data
        self.states = []
        self.actions = []
        self.probs = {}
        self.rewards = {}
        self.gamma = 0.0
        self.epsilon = 0.0

        # Solution
        self.optimal_utils = []
        self.best_action_inds = []


    def read_mdp_file(self, fname="mdpinput.txt"):
        """ Reads in mdp from file """
        with open(fname, 'r') as infile:
            csvreader = csv.reader(infile)
            lines = []
            for line in csvreader:
                lines.append(line)

            # 1) Grab the States from the 2nd line (index 1)
            for char_val in lines[1]:
                self.states.append(char_val.strip())

            # 2) Grab the Actions from the 4th line (index 3)
            for char_val in lines[3]:
                a_str = char_val.strip()
                num_s = len(self.states)
                self.actions.append(a_str)
                self.probs[a_str] = [[0.0 for i in range(num_s)] for j in range(num_s)]

            # 3) Grab the Transition models starting on 8th line (index 7)
            line_section_idx = 7
            for a_idx, action in enumerate(self.actions):
                for i in range(0, 3):
                    line_idx = line_section_idx + a_idx*4 + i
                    for j in range(0, 3):
                        self.probs[action][i][j] = float(lines[line_idx][j])


            # 4) Grab the Rewards (assumes in correct order like example)
            line_section_idx = line_section_idx + len(self.actions)*4
            line_idx = line_section_idx
            for state in self.states:
                self.rewards[state] = {}
                for action in self.actions:
                    self.rewards[state][action] = float(lines[line_idx][-1])
                    line_idx += 1

            # 5) Grab the discount factor (gamma)
            line_section_idx = line_idx + 1
            self.gamma = float(lines[line_section_idx][-1])

            # 6) Grab the tolerance (epsilon)
            line_section_idx += 2
            self.epsilon = float(lines[line_section_idx][-1])


    def run(self):
        """ Obtain Optimal Policy """
        # 1) Run Value Iteration
        utils = self.value_iteration()

        # 2) Obtain the Optimal Policy given these values
        best_action_inds = self.compute_policy(utils)

        self.optimal_utils = utils
        self.best_action_inds = best_action_inds


    def value_iteration(self):
        """ Value Iteration Algorithm """
        # Init local variables
        # utils = [0.0] * len(self.states)
        utils_ = [0.0] * len(self.states)

        # Repeat
        while 1:
            # Start the next iteration, copy util over and reset delta
            utils = copy.deepcopy(utils_)
            delta = 0.0

            # loop over each state, computing next utilities
            for idx_s, state in enumerate(self.states):
                # 1) Compute U'[s]
                curr_best_util = -math.inf
                for _, action in enumerate(self.actions):
                    curr_util = 0.0
                    for idx_s_, _ in enumerate(self.states):
                        curr_util += self.probs[action][idx_s][idx_s_] * utils[idx_s_]
                    curr_util = curr_util * self.gamma + self.rewards[state][action]
                    if curr_util > curr_best_util:
                        curr_best_util = curr_util
                utils_[idx_s] = curr_best_util

                # 2) Update delta if difference is larger
                if abs(utils_[idx_s] - utils[idx_s]) > delta:
                    delta = abs(utils_[idx_s] - utils[idx_s])

            # Until error threshold is met
            if delta < self.epsilon * (1 - self.gamma) / self.gamma:
                return utils



    def compute_policy(self, utils):
        """ Computes optimal policy from utility values after value iteration is done """
        best_action_inds = []

        for idx_s, state in enumerate(self.states):
            curr_best_util = -math.inf
            curr_best_action_idx = 0
            for idx_a, action in enumerate(self.actions):
                curr_util = 0.0
                for idx_s_, _ in enumerate(self.states):
                    curr_util += self.probs[action][idx_s][idx_s_] * utils[idx_s_]
                curr_util = curr_util * self.gamma + self.rewards[state][action]
                if curr_util > curr_best_util:
                    curr_best_util = curr_util
                    curr_best_action_idx = idx_a
            best_action_inds.append(curr_best_action_idx)

        return best_action_inds


    def write_output(self, fname="policy.txt"):
        """ Write formatted output of to output file """
         # 1) Open the file for writing
        f = open(fname, "w")

        # 2) Construct the output string
        out_str = ""
        for idx_s, state in enumerate(self.states):
            out_str += state + ": "
            out_str += self.actions[self.best_action_inds[idx_s]]
            out_str += " (" + "{:.2f}".format(self.optimal_utils[idx_s]) + ")"

            if idx_s < (len(self.states) - 1):
                out_str += "\n"

        # 3) Write to the output file and then close it
        f.write(out_str)
        f.close()


def main():
    """ main function for calling mdp value iteration algorithm """
    my_mdp = MDP()

    # 1) Read in the input file
    my_mdp.read_mdp_file()

    # 2) Perform Value Iteration to compute the optimal policy
    my_mdp.run()

    # 3) Write Output
    my_mdp.write_output()



if __name__ == "__main__":
    main()

File: Project5/Solutions/cpp/mdp.cpp
#include "mdp.h"

using std::cout;
using std::endl;

/*
*   MDP Constructor
*/
MDP::MDP() {
}

//////////////////////////////////////////////////////////////////////////////
//                                                                          //
//                                                                          //
//                              User Functions                              //
//                                                                          //
//                                                                          //
//////////////////////////////////////////////////////////////////////////////

/*
*   readInput:  Reads in inFilename and populates mdp data
*
*       Returns 0 on success, 1 on failure
*/
int MDP::readInput() {
    std::ifstream inFile(inFilename);
    std::string line, val;
    double probVal;

    // Read in the data
    if (inFile.is_open()) {
        // 1) States
        getline(inFile, line);
        if (getline(inFile, line)) {
            while (line.length() > 0) {
                std::size_t first_comma = line.find_first_of(',');
                if (first_comma == string::npos) {
                    states.push_back(line);
                    nameToState[line] = numStates;
                    numStates++;
                    break;
                }
                states.push_back(line.substr(0, first_comma));
                nameToState[line.substr(0, first_comma)] = numStates;
                line = line.substr(first_comma + 2);
                numStates++;
            }
        } else { return 1; }

        // 2) Actions
        getline(inFile, line);
        if (getline(inFile, line)) {
            while (line.length() > 0) {
                std::size_t first_comma = line.find_first_of(',');
                if (first_comma == string::npos) {
                    actions.push_back(line);
                    nameToAction[line] = numActions;
                    numActions++;
                    break;
                }
                actions.push_back(line.substr(0, first_comma));
                nameToAction[line.substr(0, first_comma)] = numActions;
                line = line.substr(first_comma + 2);
                numActions++;
            }

        } else { return 1; }

        // 2.5) Set up P_ Variable
        for (int i=0; i < numActions; i++) {
            P_.push_back(vector<vector<double>>());
            for (int j=0; j < numStates; j++) {
                P_[i].push_back(vector<double>());
            }
        }

        // 3) Transition Model
        getline(inFile, line);
        getline(inFile, line);
        for (int i=0; i < numActions; i++) {
            getline(inFile, line);
            for (int j=0; j < numStates; j++) {
                if (getline(inFile, line)) {
                    while (line.length() > 0) {
                        std::size_t first_comma = line.find_first_of(',');
                        if (first_comma == string::npos) {
                            probVal = std::stod(line);
                            P_[i][j].push_back(probVal);
                            break;
                        }
                        probVal = std::stod(line.substr(0, first_comma));
                        P_[i][j].push_back(probVal);
                        line = line.substr(first_comma + 2);
                    }
                } else { return 1; }
            }
        }

        // 3.5) Set up R_ Vector
        for (int i=0; i < numStates; i++) {
            R_.push_back(vector<double>());
            for (int j=0; j < numActions; j++) {
                R_[i].push_back(0);
            }
        }

        // 4) Rewards
        getline(inFile, line);
        for (int i=0; i < numStates; i++) {
            for (int j=0; j < numActions; j++) {
                if (getline(inFile, line)) {
                    std::size_t first_comma = line.find_first_of(',');
                    std::size_t last_comma = line.find_last_of(',');

                    string theState = line.substr(0, first_comma);
                    string theAction = line.substr(first_comma+2,
                                                   last_comma-first_comma-2);
                    string theValue = line.substr(last_comma+2);

                    int s = nameToState[theState];
                    int a = nameToAction[theAction];
                    double zeValue = std::stod(theValue);

                    R_[s][a] = zeValue;
                } else { return 1; }
            }
        }

        // 5) Discount Factor
        getline(inFile, line);
        if (getline(inFile, line)) {
            gamma = std::stod(line);
        } else { return 1; }

        // 6) Epsilon (tolerance)
        getline(inFile, line);
        if (getline(inFile, line)) {
            eps = std::stod(line);
        } else { return 1; }
    } else {
        std::cout << "Error Opening " << inFilename << std::endl;
        std::cout << "Double Check Spelling, its presence, etc." << std::endl;
        return 1;
    }

    inFile.close();

    return 0;
}


/*
*   run
*/
void MDP::run() {
    // 1) Obtain the Utilities using valueIteration
    U_final = valueIteration();

    // 2) Obtain the Policy
    for (int s=0; s < numStates; s++) {
        double maxValue = negInf;
        int maxAction = 0;

        for (int a=0; a < numActions; a++) {
            double currValue = 0;
            for (int s_ = 0; s_ < numStates; s_++) {
                currValue += P(s_, s, a) * U_final[s_];
            }
            currValue = currValue * gamma + R(s, a);

            if (currValue > maxValue) {
                maxValue = currValue;
                maxAction = a;
            }
        }

        policy.push_back(maxAction);
    }
}


/*
*   writeOutput
*/
int MDP::writeOutput() {
    std::ofstream outFile(outFilename);

    outFile << "% Format: State: Action (Value)" << endl;
    for (int s=0; s < numStates; s++) {
        outFile << states[s] << ": ";
        outFile << actions[policy[s]] << " ";
        outFile << "(" << U_final[s] << ")" << endl;
    }

    outFile.close();
    return 0;
}

//////////////////////////////////////////////////////////////////////////////
//                                                                          //
//                                                                          //
//                       Value Iteration Functions                          //
//                                                                          //
//                                                                          //
//////////////////////////////////////////////////////////////////////////////

/*
*   valueIteration  (R&N Figure 17.4)
*/
vector<double> MDP::valueIteration() {
    // local variables
    vector<double> U(numStates, 0);
    vector<double> U_(numStates, 0);
    double delta = 0;

    while (1) {
        U = U_;
        delta = 0;
        for (int s=0; s < numStates; s++) {
            double maxValue = negInf;
            for (int a=0; a < numActions; a++) {
                double nextVal = 0;
                for (int s_=0; s_ < numStates; s_++) {
                    nextVal += P(s_, s, a) * U[s_];
                }
                nextVal = nextVal * gamma + R(s, a);

                if (nextVal > maxValue) {
                    maxValue = nextVal;
                }
            }
            U_[s] = maxValue;

            if (fabs(U_[s] - U[s]) > delta) {
                delta = fabs(U_[s] - U[s]);
            }
        }

        if (delta < eps * (1 - gamma) / gamma) {
            break;
        }
    }

    return U;
}


//////////////////////////////////////////////////////////////////////////////
//                                                                          //
//                                                                          //
//                           Helper Functions                               //
//                                                                          //
//                                                                          //
//////////////////////////////////////////////////////////////////////////////






















